### PR TITLE
ocamlPackages.melange: init at 4.0.1-52 + 4.0.0-51 + 4.0.0-414

### PR DESCRIPTION
**Pull request by vog**
Add package ocamlPackages.melange.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2505.section.md [hunk 1]
```diff
@@ -12,6 +12,8 @@
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
+- [Melange](https://melange.re/), the toolchain to produce JavaScript from Reason/OCaml.
```
Comment (Lassulus) [inline nixos/doc/manual/release-notes/rl-2505.section.md:15]: for now this is just a package, so I would leave the release note out

File: pkgs/development/tools/ocaml/melange/default.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{
```
Comment (Lassulus) [inline pkgs/development/tools/ocaml/melange/default.nix:1]: not sure if this file could go into pkgs/by-name. generally we try to move everything there, but there could be exceptions for language specific packages that I'm not aware of

Comment (vog) [reply pkgs/development/tools/ocaml/melange/default.nix:1]: I tried to make it as similar as possible to the js_of_ocaml package, which serves a very similar purpose.

File: nixos/doc/manual/release-notes/rl-2505.section.md (hunk 1 above)
Comment (vog) [reply nixos/doc/manual/release-notes/rl-2505.section.md]: Ok, I just removed that entry from my commit.

File: pkgs/development/tools/ocaml/melange/default.nix (hunk 2 above)
Comment (vog) [reply pkgs/development/tools/ocaml/melange/default.nix:1]: Note that there is also a kind-of naming conflict with the existing "pkgs.melange" package (completely unrelated package), which is not an actual conflict as this one goes to "pkgs.ocamlPackages.melange".

Comment (vog) [reply pkgs/development/tools/ocaml/melange/default.nix:1]: Thinking more of this, should I rename this package to "melange-re" and move it `pkgs/by-name/me/melange-re`?

If so, would it be okay to still be named `ocamlPackages.melange` in the Nix code, or should it be renamed to `ocamlPackages.melange-re` as well?

Comment (Lassulus) [reply pkgs/development/tools/ocaml/melange/default.nix:1]: I think the later sounds like a good idea, but I'm not an expert in the ocaml ecosystem :)

**Review by anmonteiro — COMMENTED** [review]
this can be done for more OCaml versions rather than just 5.2 https://github.com/nix-ocaml/nix-overlays/blob/master/ocaml/melange/default.nix

File: pkgs/development/tools/ocaml/melange/default.nix (hunk 2 above)
Comment (vog) [reply pkgs/development/tools/ocaml/melange/default.nix:1]: Ok, will do.

Comment (vog) [reply pkgs/development/tools/ocaml/melange/default.nix:1]: Well, melange also works with OCaml 4.14, what should we do about that?

**Comment by vog** [comment]
> this can be done for more OCaml versions rather than just 5.2 https://github.com/nix-ocaml/nix-overlays/blob/master/ocaml/melange/default.nix

Ok, I'll take care of that.

File: pkgs/development/tools/ocaml/melange/default.nix (hunk 2 above)
Comment (anmonteiro) [reply pkgs/development/tools/ocaml/melange/default.nix:1]: I think you've already added it to the right place, but yeah let's add the 4.14 version too

**Comment by vog** [comment]
I believe the latest version of my pull requests addresses all discussed issues.

@anmonteiro @Lassulus Is there anything else that I should take care of?

**Review by ulrikstrid — APPROVED** [review]

**Pull request merged**